lib/heex_formatter/phases/tokenizer.ex

```diff
@@ -27,46 +31,160 @@ defmodule HeexFormatter.Phases.Tokenizer do
-    string = List.to_string(text)
-    {tokens, _} = HTMLTokenizer.tokenize(string, "nofile", 0, [], [], :text)
-    {tokens |> Enum.reject(&line_break?/1) |> Enum.reverse(), acc}
+  defp do_tokenize({:text, _line, _column, text}, {tokens, cont}) do
+    text
+    |> List.to_string()
+    |> HTMLTokenizer.tokenize("nofile", 0, [], tokens, cont)
   end
 
   @eex_expr [:start_expr, :expr, :end_expr, :middle_expr]
 
-  defp tokenize({type, line, column, opt, expr}, acc) when type in @eex_expr do
+  defp do_tokenize({type, _, _, opt, expr}, {tokens, :script = cont})
+       when type in @eex_expr do
+    expr = expr |> List.to_string() |> String.trim()
+
+    eex_tag =
+      if List.to_string(opt) == "=" do
+        "<%= #{expr} %>"
+      else
+        "<% #{expr} %>"
+      end
+
+    # Get the last token which is the one the we want to append this eex_tag with.
+    [{:text, text, meta} | rest] = tokens
+
+    {[{:text, text <> eex_tag, meta} | rest], cont}
+  end
+
+  # When the mode is comment, we concatenate the iex expression with the previous
+  # text so that we don't need to worry about it when formatting the HTML.
+  defp do_tokenize({type, _, _, opt, expr}, {tokens, {:comment, _, _} = cont})
+       when type in @eex_expr do
+    expr = expr |> List.to_string() |> String.trim()
+
+    eex_tag =
+      if List.to_string(opt) == "=" do
+        "<%= #{expr} %>"
+      else
+        "<% #{expr} %>"
+      end
+
+    # Get the last token which is the one the we want to append this eex_tag with.
+    [{:text, text, meta} | rest] = tokens
+
+    {[{:text, text <> eex_tag, meta} | rest], cont}
+  end
+
+  defp do_tokenize({type, line, column, opt, expr}, {tokens, cont}) when type in @eex_expr do
     render = List.to_string(opt)
     expr = expr |> List.to_string() |> String.trim()
     block? = String.ends_with?(expr, "do") || String.ends_with?(expr, "->")
     meta = %{column: column, line: line, block?: block?}
 
-    {type, tag} =
+    token =
       if render == "=" do
-        {:eex_tag_render, "<%= #{expr} %>"}
+        {:eex_tag_render, "<%= #{expr} %>", meta}
       else
-        {:eex_tag, "<% #{expr} %>"}
+        {:eex_tag, "<% #{expr} %>", meta}
       end
 
-    {[{type, tag, meta}], acc}
+    {[token | tokens], cont}
   end
 
-  defp tokenize(_node, acc) do
-    {[], acc}
+  defp do_tokenize(_node, acc) do
+    acc
   end
 
-  defp line_break?({:text, text, _meta}) do
+  @doc """
+  Finalize the Tokenization started by `Tokenizer.tokenize/1`.
+
+  This function is going to do some clean up and normalization to help the
+  Formatter module. Here are the final steps it does:
+
+  * Concatenate text from tags: script, tag, style, code, pre  and HTML comments
+    so that we don't we just append as it is in the Formatter.
+
+  * Ignore line breaks that comes from HTMLTokenizer. That is because the Formatter
+    already know to handle line breanks.
+  """
+  def finalize(tokens) do
```

Comment:
Ok, we might don't need it at all since we could handle this when tokenizing. This PR is supposed to help with that:

https://github.com/phoenixframework/phoenix_live_view/pull/1847